src/coreclr/src/vm/reflectioninvocation.cpp

```diff
@@ -1760,35 +1583,41 @@ void QCALLTYPE ReflectionInvocation::CompileMethod(MethodDesc * pMD)
 }
 
 // This method triggers the class constructor for a give type
-FCIMPL1(void, ReflectionInvocation::RunClassConstructor, ReflectClassBaseObject *pTypeUNSAFE)
+void QCALLTYPE ReflectionInvocation::RunClassConstructor(
+    QCall::TypeHandle pTypeHandle,
+    BOOL fPreciseCctorsOnly)
 {
-    FCALL_CONTRACT;
-
-    REFLECTCLASSBASEREF refType = (REFLECTCLASSBASEREF)ObjectToOBJECTREF(pTypeUNSAFE);
-
-    if (refType == NULL)
-        FCThrowArgumentVoidEx(kArgumentException, NULL, W("InvalidOperation_HandleIsNotInitialized"));
-
-    TypeHandle typeHnd = refType->GetType();
-    if (typeHnd.IsTypeDesc())
-        return;
+    QCALL_CONTRACT;
 
-    MethodTable *pMT = typeHnd.AsMethodTable();
+    BEGIN_QCALL;
 
-    Assembly *pAssem = pMT->GetAssembly();
+    TypeHandle typeHnd = pTypeHandle.AsTypeHandle();
+    if (typeHnd.IsNull())
```

Comment:
Actually, `IsClassInited` is not correct fast-path condition for the `fPreciseCctorsOnly` case. The `fPreciseCctorsOnly` is expected to run all constructors through the inheritance chain.

It make me wonder whether it makes sense to have one QCall for both cases. It may be better to have two different QCalls.

Reply:
Resolved this offline - we'll make the unmanaged side responsible for running cctors, so this becomes moot.